doc/lexical_envs.rst

```diff
@@ -0,0 +1,265 @@
+  brackets, and (2) the declaration note for ``int a``.
+
+  The "node_env" of (1) is whatever env one of its parent nodes has created.
+  (2) lives in the scope of (1), so "node_env" for (1) would be the environment
+  that (2) has created.
+
+* "children_env" returns the same as "node_env" if the node does not create an
+  environment, or the environment it has created otherwise.
+
+  To continue on the previous example: the "children_env" of (1) is the
+  environment this node has created (this is (2)'s "node_env"). However
+  the "children_env" of (2) is (1)'s "children_env", as (2) does not create
+  environments, "node_env" and "children_env" are the same for it.
+
+
+Env specs
+=========
+
+An env spec is a sequence of actions ("env actions") whose goal is to create
+the mesh of lexical environments that is the most appropriate to model the
+analyzed language's scoping rules, so that properties can easily query them to
+implement semantic analysis.
+
+Note that unlike the properties DSL, which uses a pure functional paradigm
+(there are exceptions, but by default properties have no side effect), env spec
+are essentially imperative: actions are executed in sequence, they modify the
+state of the nodes that are traversed, and thus change the behavior of future
+property calls.
+
+Each env action performs an elementary action. For instance:
+
+* Set the "current" lexical environment (more about what this means in the next
+  section).
+
+* Create a lexical environment.
+
+* Add a symbol/node association in the current environment.
+
+
+Populate Lexical Env (PLE)
+==========================
+
+At runtime, env actions are executed during a pass called "Populate Lexical
+Env" (PLE): env actions are executed for all nodes in one analysis unit in a
+depth-first-search fashion.
+
+For each node, the PLE pass selects the list of env actions to execute
+depending on the node type. Also, env actions execution operates on a "current
+environment", which env actions can mutate or replace. Given one node, PLE runs
+through the following steps:
+
+* Inherit the "current environment" from the PLE run on the parent node.
+* Set the node's "node_env/children_env" to this current env.
+* Execute the "pre" env actions.
+* If the current env has changed, update the node's "node_env/children_env".
+* Run PLE on the node children.
+* execute the "post" env actions.
+
+Regarding env actions selection: by default env specs are inherited. For
+instance, if node A has an env spec, node B derives from A and does not have an
+env spec, A's env spec applies to B. And if node C also derives from and but
+defines an env spec, A's env spec does not apply to C.
+
+Users can manually run PLE, but they don't, it is automatically executed when
+needed, and is executed at most once per analysis unit (until the unit is
+reparsed).
+
+
+Basic example
+=============
+
+The following simple example will hopefully clarify the above explanations.
```

Comment:
Remove hopefully: "Let's use the following example to show how the above would work in a concrete use case" or sth